### PR TITLE
CHT calculations for full nodes

### LINKED ISSUE #3458: Canonical Hash Trie calculations for full node
The canonical hash trie is used in LES for getting relatively trusted proofs of historical header hashes and values. Full nodes serving the LES protocol should generate these.

Open questions:
  - Should they go into the state column or a separate column?
  - Should they be subject to the same pruning history as the rest of the state?

### PR DESCRIPTION
**Pull request by rphmeier**
Closes #3458 (for now)

For now, Tries are always calculated on-demand. This will likely be changed in the future to maintain an LRU-Cache of recently-generated CHTs.

Also adds some more tests and documentation.

### CODE REVIEW

**Comment by rphmeier** [comment]
Note: this PR will be the first containing a divergence from LES as-specified. Perhaps we should choose another protocol identifier? `plp` for Parity Light Protocol?

**Comment by arkpar** [comment]
Does it break compatibility with go-ethereum implementation?

**Comment by rphmeier** [comment]
yes, although there's no server/client capability exposed yet.

File: ethcore/light/src/provider.rs [hunk 1]
```diff
@@ -226,11 +226,76 @@ impl<T: ProvingBlockChainClient + ?Sized> Provider for T {
...
+							&*::rlp::encode(&i),
+							&*::rlp::encode(&hdr.hash()),
+						).expect("fresh in-memory database is infallible; qed");
+
+						if i == req.block_number { needed_hdr = Some(hdr) }
+					}
+				}
+			}
+		}
+		let needed_hdr = needed_hdr.expect("`needed_hdr` always set in loop, number checked before; qed");
```
Comment (arkpar) [inline ethcore/light/src/provider.rs]: It is not really clear that `needed_hdr` is always set. Is `req.block_number` guaranteed to be within the range?

Comment (rphmeier) [inline ethcore/light/src/provider.rs]: it's guaranteed by the check that `req.cht_number` is equal to the calculated CHT number based on the block number in the request. this means that the function has already returned in the case that the requested number wouldn't fall within the range of blocks we iterate.

Comment (rphmeier) [inline ethcore/light/src/provider.rs]: Additionally, if we don't have all the blocks in that range, we also return early. The CHT number should probably be omitted in the future _because_ it is derivable from a block number and only serves to give misbehaving peers a method of misbehavior.

**Pull request merged**